src/thunderbots/software/radio_communication/mrf/dongle.cpp

```diff
@@ -0,0 +1,610 @@
+    // While doing so, discover which interface number is used for the radio and
+    // which alternate settings are for configuration-setting and normal
+    // operation.
+    {
+        const libusb_config_descriptor &desc =
+            device.configuration_descriptor_by_value(1);
+        for (int i = 0; i < desc.bNumInterfaces; ++i)
+        {
+            const libusb_interface &intf = desc.interface[i];
+            if (intf.num_altsetting && intf.altsetting[0].bInterfaceClass == 0xFF &&
+                intf.altsetting[1].bInterfaceSubClass == MRF::SUBCLASS)
+            {
+                radio_interface = i;
+                for (int j = 0; j < intf.num_altsetting; ++j)
+                {
+                    const libusb_interface_descriptor &as = intf.altsetting[j];
+                    if (as.bInterfaceClass == 0xFF &&
+                        as.bInterfaceSubClass == MRF::SUBCLASS)
+                    {
+                        if (as.bInterfaceProtocol == MRF::PROTOCOL_OFF)
+                        {
+                            configuration_altsetting = j;
+                        }
+                        else if (as.bInterfaceProtocol == MRF::PROTOCOL_NORMAL)
+                        {
+                            normal_altsetting = j;
+                        }
+                    }
+                }
+                break;
+            }
+        }
+        if (radio_interface < 0 || configuration_altsetting < 0 || normal_altsetting < 0)
+        {
+            throw std::runtime_error(
+                "Wrong USB descriptors (is your dongle firmware or your "
+                "software out of date or mismatched across branches?).");
+        }
+    }
+
+    // Move the dongle into configuration 1 (it will nearly always already be
+    // there).
+    if (device.get_configuration() != 1)
+    {
+        device.set_configuration(1);
+    }
+
+    // Claim the radio interface.
+    interface_claimer.reset(new USB::InterfaceClaimer(device, radio_interface));
+
+    // Switch to configuration mode and configure the radio parameters.
+    device.set_interface_alt_setting(radio_interface, configuration_altsetting);
+    {
+        unsigned int config = 0U;
+        {
+            const char *config_string = std::getenv("MRF_CONFIG");
+            if (config_string)
+            {
+                int i = std::stoi(config_string, nullptr, 0);
+                if (i < 0 || static_cast<std::size_t>(i) >=
+                                 sizeof(DEFAULT_CONFIGS) / sizeof(*DEFAULT_CONFIGS))
+                {
+                    throw std::out_of_range(
+                        "Config index must be between 0 and number of configs "
+                        "- 1.");
+                }
+                config = static_cast<unsigned int>(i);
+            }
+        }
+        channel_ = DEFAULT_CONFIGS[config].channel;
+        {
+            const char *channel_string = std::getenv("MRF_CHANNEL");
```

Comment:
Should we be setting this somewhere? Like in a launch file or something? Same goes for the other environment variables below?

Reply:
In the majority of use cases these variables do not need to be touched (in fact I don't remember ever changing any of these aside from the channel). The channel can be changed so we can have 2 dongles communicate with different subsets of robots (should probably make a ticket for this)